README.md
[![Actions Status](https://github.com/zakame/hashids.pm/workflows/Test%20on%20latest%20supported%20Perls/badge.svg)](https://github.com/zakame/hashids.pm/actions) [![Coverage Status](https://img.shields.io/coveralls/zakame/hashids.pm/master.svg?style=flat)](https://coveralls.io/r/zakame/hashids.pm?branch=master) [![MetaCPAN Release](https://badge.fury.io/pl/Hashids.svg)](https://metacpan.org/release/Hashids) [![Build Status](https://img.shields.io/appveyor/ci/zakame/hashids-pm/master.svg?logo=appveyor)](https://ci.appveyor.com/project/zakame/hashids-pm/branch/master)
# NAME

Hashids - generate short hashes from numbers

# SYNOPSIS

    use Hashids;
    my $hashids = Hashids->new('this is my salt');

    # encrypt a single number
    my $hash = $hashids->encode(123);          # 'YDx'
    my $number = $hashids->decode('YDx');      # 123

    # or a list
    $hash = $hashids->encode(1, 2, 3);         # 'laHquq'
    my @numbers = $hashids->decode('laHquq');  # (1, 2, 3)

    # also get results in an arrayref
    my $numbers = $hashids->decode('laHquq');  # [1, 2, 3]

# DESCRIPTION

This is a port of the Hashids JavaScript library for Perl.

Hashids was designed for use in URL shortening, tracking stuff,
validating accounts or making pages private (through abstraction.)
Instead of showing items as `1`, `2`, or `3`, you could show them as
`b9iLXiAa`, `EATedTBy`, and `Aaco9cy5`.  Hashes depend on your salt
value.

**IMPORTANT**: This implementation follows the v1.0.0 API release of
hashids.js.  An older API of hashids.js (v0.1.4) can be found in Hashids
version 0.08 and earlier releases; if you have code that depends on this
API version, please use a tool like [Carton](https://metacpan.org/pod/Carton) to pin your Hashids
install to the older version.

This implementation is also compatible with the v0.3.x hashids.js API.

# METHODS

## new

    my $hashids = Hashids->new();

Make a new Hashids object.  This constructor accepts a few options:

    my $hashids = Hashids->new(
        salt          => 'this is my salt',
        alphabet      => 'abcdefghijklmnop',
        minHashLength => 8
    );

- salt

    Salt string, this should be unique per Hashids object.  Must be either
    as long or shorter than the alphabet length, as a longer salt string
    than the alphabet introduces false collisions.

- alphabet

    Alphabet set to use.  This is optional as Hashids comes with a default
    set suitable for URL shortening.  Should you choose to supply a custom
    alphabet, make sure that it is at least 16 characters long, has no
    spaces, and only has unique characters.

- minHashLength

    Minimum hash length.  Use this to control how long the generated hash
    string should be.

You can also construct with just a single argument for the salt, leaving
the alphabet and minHashLength at their defaults:

    my $hashids = Hashids->new('this is my salt');

## encode

    my $hash = $hashids->encode($x, [$y, $z, ...]);

Encode a single number (or a list of numbers) into a hash string.

## encrypt

Alias for ["encode"](#encode), for compatibility with v0.3.x hashids.js API.

## encode\_hex

    my $hash = $hashids->encode_hex('deadbeef');

Encode a hex string into a hash string.

## decode

    my $number = $hashids->decode($hash);

Decode a hash string into its number (or numbers.)  Returns either a
simple scalar if it is a single number, an arrayref of numbers if it
decrypted a set, or `undef` if given bad input.  Use ["ref" in perlfunc](https://metacpan.org/pod/perlfunc#ref) on
the result to ensure proper usage.

You can also retrieve the result as a proper list by assigning it to an
array variable, by doing so you will always get a list of one or more
numbers that are decrypted from the hash, or the empty list if none were
found:

    my @numbers = $hashids->decode($hash);

## decrypt

Alias for this ["decode"](#decode), for compatibility with v0.3.x hashids.js API.

## decode\_hex

    my $hex_string = $hashids->decode_hex($hash);

Opposite of ["encode\_hex"](#encode_hex).  Unlike ["decode"](#decode), this will always return
a string, including the empty string if the hash is invalid.

# SEE ALSO

[Hashids](http://www.hashids.org)

# LICENSE

The MIT License (MIT)

Copyright (C) Zak B. Elep.

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

# AUTHOR

Zak B. Elep <zakame@cpan.org>

Original Hashids JavaScript library written by [Ivan
Akimov](http://twitter.com/ivanakimov)

# THANKS

Props to [Jofell Gallardo](http://twitter.com/jofell) for pointing this
excellent project to me in the first place.

Many thanks to [C. A. Church](https://github.com/thisdroneeatspeople)
and [Troy Morehouse](https://github.com/tmorehouse) for their fixes and
updates.


Changes
Revision history for Perl extension Hashids

{{$NEXT}}

1.001013 2020-12-17T07:40:44Z

    - Maintenance release, no functional changes
    - Switch from Travis CI to GitHub Actions

1.001012 2020-01-25T02:25:44Z

    - Fix bignum pragma check in misc test

1.001011 2020-01-24T03:52:21Z

    - Hashids::Util: fix possible infinite loop in to_alphabet() when
      bignum pragma is loaded, thanks to mla (closes #14)
    - Adjust Travis and AppVeyor CI configurations  

1.001010 2018-10-10T05:36:22Z

    - Fix output comment in synopsis, thanks to Dai Okabayashi
    - Small update on the hex methods, needing a bump of Math::BigInt

1.001001 2018-03-05T11:40:03Z

    - Use "[0-9]" instead of "\d" in number matches to prevent possible
      Unicode digit issues
    - Fix example for Hashids::Util::any()

1.001000 2017-11-04T09:00:59Z

    - Align with v1.1.x of Hashids.js
    - Hashids: use Class::Method::Modifiers::around() for BUILDARGS
    - Hashids::Util: make consistent_shuffle() to always return lists
    - Add constraint on salt length to be less than or equal to alphabet
      length (see also https://github.com/ivanakimov/hashids.js/issues/37)
    - Hashids::Util: add any() to avoid using boolean grep()
    - Add more tests adapted from Hashids.js
    - Various other small fixes suggested by Perl::Critic and kritika.io

1.000005 2017-06-24T13:42:15Z

    - Hashids::Util: ensure we carry a bignum() upon reduce() in
      from_alphabet(), fixes #11
    - Add appveyor CI (no badge yet though, needs fix in Minilla)

1.000004 2017-06-23T06:40:54Z

    - Reduce calls to Hashids::Util::bignum() to further simplify code

1.000003 2017-06-23T05:51:41Z

    - Add test for encoding lists with set minHashLength
    - Move private functions into Hashids::Util module
    - Rename _hash/_unhash to to_alphabet/from_alphabet for
      consistency with upstream JS methods
    - Use Horner's method on from_alphabet

1.000002 2015-01-22T09:01:42Z

    - Switched calculations to use Math::BigInt, thanks to Troy
      Morehouse (closes #6)

1.000001 2014-12-04T06:54:57Z

    - Fix issue where some hashes could not be decoded, thanks to
      C. A. Church (closes #7)

1.000000 2014-09-10T13:00:52Z

    - Bump to 1.0.0 hashids.js API, closes #5
    - Add encode_hex() and decode_hex() methods
    - Add Perl 5.20 to travis build
    - Add coveralls test coverage reporting
    - Add more tests

0.08 2013-10-20T10:58:30Z

    - Check intial arguments sooner
    - Adjust travis build (use Perl 5.14 to 5.18)
    - Add Travisa badge in README

0.07 2013-08-30T11:42:51Z

    - Fix 'map ord reverse split...' for perls < 5.18

0.06 2013-08-30T11:14:06Z

    - Add note regarding algorithm versions
    - Croak instead of die
    - No more using List::MoreUtils and Scalar::Util
    - encrypt() accepts counting numbers only
    - Refactor some more

0.05 2013-08-26T10:40:41Z

    - Adjust deps
    - A lot more refactoring

0.04 2013-08-22T18:10:56Z

    - Fix a bug in decrypt() when subclassing
    - Let decrypt() emit sane output in when in list context, and be
      symmetric to encrypt()
    - Remove Test::Deep, use Test::More::is_deeply() instead

0.03 2013-08-20T15:57:59Z

    - Let the construct accept salt string only for consistency with
      other Hashids implementations
    - Refactor a really little bit :)

0.02 2013-08-20T10:01:07Z

    - Forgot salt key in SYNOPSIS

0.01 2013-08-20T08:41:48Z

    - initial release, a straight port of hashids.js for Perl.  Expect
      updates soon :)



t/lib/SubClassTest.pm
package SubClassTest;
use Moo;
extends 'Hashids';

has extra_number => ( is => 'ro', required => 1 );
has '+salt' => ( default => sub {'I want peppers'} );

around encode => sub {
    my ( $orig, $self, @args ) = @_;
    $self->$orig( $self->extra_number, @args );
};

1;


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Hashids',
    module_name     => 'Hashids',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,


);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

use File::Copy;

print "cp META.json MYMETA.json\n";
copy("META.json","MYMETA.json") or die "Copy failed(META.json): $!";

if (-f 'META.yml') {
    print "cp META.yml MYMETA.yml\n";
    copy("META.yml","MYMETA.yml") or die "Copy failed(META.yml): $!";
} else {
    print "There is no META.yml... You may install this module from the repository...\n";
}



lib/Hashids/Util.pm
package Hashids::Util;

use strict;
use warnings;

our ( @EXPORT_OK, %EXPORT_TAGS );

BEGIN {
    use Exporter 'import';
    @EXPORT_OK = qw(consistent_shuffle to_alphabet from_alphabet any bignum);
    %EXPORT_TAGS = ( all => \@EXPORT_OK );
}

use List::Util 'reduce';
use Math::BigInt;

use namespace::clean -except => [qw(import)];

sub consistent_shuffle {
    my ( $alphabet, $salt ) = @_;

    return ('') unless $alphabet;

    my @alphabet
        = ref $alphabet eq 'ARRAY' ? @$alphabet : split // => $alphabet;
    return @alphabet unless $salt;
    my @salt = ref $salt eq 'ARRAY' ? @$salt : split //, $salt;

    for ( my ( $i, $v, $p ) = ( $#alphabet, 0, 0 ); $i > 0; $i--, $v++ ) {
        $p += my $int = ord $salt[ $v %= @salt ];
        my $j = ( $int + $v + $p ) % $i;

        @alphabet[ $j, $i ] = @alphabet[ $i, $j ];
    }

    @alphabet;
}

sub to_alphabet {
    my ( $num, $alphabet ) = @_;

    my $hash = '';
    my @alphabet
        = ref $alphabet eq 'ARRAY' ? @$alphabet : split // => $alphabet;

    $num = bignum($num);
    do {
        $hash = $alphabet[ $num % @alphabet ] . $hash;
        $num = int($num / @alphabet);
    } while ( $num != 0 );

    $hash;
}

sub from_alphabet {
    my ( $hash, $alphabet ) = @_;

    my @alphabet
        = ref $alphabet eq 'ARRAY' ? @$alphabet : split // => $alphabet;

    my $num = reduce { bignum($a) * @alphabet + $b }
    map { index join( '' => @alphabet ), $_ } split // => $hash;

    "$num";
}

sub any (&@) {                 ## no critic (ProhibitSubroutinePrototypes)
    my $f = shift;
    for (@_) {
        return 1 if $f->();
    }
    0;
}

sub bignum {
    my $n = Math::BigInt->bzero();
    $n->round_mode('zero');
    $n->badd("@{[shift]}");
}

1;
__END__

=encoding utf-8

=for stopwords arrayrefs bignum

=head1 NAME

Hashids::Util - Shuffling and conversion functions for Hashids

=head1 SYNOPSIS

    use Hashids::Util
        qw( consistent_shuffle to_alphabet from_alphabet any bignum );

=head1 DESCRIPTION

C<Hashids::Util> provides utility functions for L<Hashids> for
consistent shuffling of alphabets, conversion of numbers to hash
strings, and others.  These are originally included as private
functions, but are now provided here as a separate module as they can be
general enough to be used in other code.

C<Hashids::Util> does not export any functions by default.

=head1 FUNCTIONS

=head2 consistent_shuffle

    my @shuffled_alphabet = consistent_shuffle( $alphabet, $salt );

Given an alphabet and salt, produce a shuffled alphabet.  Both alphabet
and salt can be either strings or arrayrefs of characters.

=head2 to_alphabet

    my $hash = to_alphabet( $num, $alphabet );

Produce a hash string given a number and alphabet.  The given alphabet
may be a string or an arrayref of characters.

=head2 from_alphabet

    my $num = from_alphabet( $hash, $alphabet );

Produce a number from the given hash string and alphabet.  The given
alphabet may be a string or arrayref of characters.

=head2 any

    print "At least one non-negative value"
        if any { $_ >= 0 } @list_of_numbers;

Returns a true value if any item in the given list meets the given
criterion.  Returns false otherwise.  Adapted from
L<List::MoreUtils::PP>.

=head2 bignum

   my $bignum = bignum( '1_152_921_504_606_846_976' );

Construct a L<Math::BigInt> scalar from the given number.

=head1 SEE ALSO

L<Hashids>

=head1 AUTHOR

Zak B. Elep E<lt>zakame@cpan.orgE<gt>

=cut


lib/Hashids.pm
package Hashids;

use Carp 'croak';
use POSIX 'ceil';
use Hashids::Util ':all';
use Moo;
use namespace::clean;

our $VERSION = "1.001013";

has salt => ( is => 'ro', default => '' );

has minHashLength => (
    is  => 'ro',
    isa => sub {
        croak "$_[0] must be a positive number" unless $_[0] =~ /^[0-9]+$/;
    },
    default => 0
);

has alphabet => (
    is  => 'rwp',
    isa => sub {
        local $_ = shift;
        croak "$_ must not have spaces" if /\s/;
        croak "$_ must contain at least 16 characters" if 16 > length;
        my %u;
        croak "$_ must contain unique characters"
            if any { $u{$_}++ } split //;
    },
    default => sub { join '' => 'a' .. 'z', 'A' .. 'Z', 1 .. 9, 0 }
);

has chars => ( is => 'rwp', init_arg => undef, default => sub { [] } );

has seps => (
    is       => 'rwp',
    init_arg => undef,
    default  => sub {
        my @seps = qw(c f h i s t u);
        [ @seps, map {uc} @seps ];
    },
);

has guards => ( is => 'rwp', init_arg => undef, default => sub { [] } );

around BUILDARGS => sub {
    my ( $orig, $class, @args ) = @_;
    unshift @args, 'salt' if @args % 2 == 1;

    $class->$orig(@args);
};

sub BUILD {
    my $self = shift;

    croak "salt must be shorter than or of equal length to alphabet"
        if length $self->salt > length $self->alphabet;

    my @alphabet = split // => $self->alphabet;
    my ( @seps, @guards );

    my $sepDiv   = 3.5;
    my $guardDiv = 12;

    # seps should contain only chars present in alphabet;
    # alphabet should not contain seps
    for my $sep ( @{ $self->seps } ) {
        push @seps, $sep if any {/$sep/} @alphabet;
        @alphabet = grep { !/$sep/ } @alphabet;
    }

    @seps = consistent_shuffle( \@seps, $self->salt );

    if ( !@seps || ( @alphabet / @seps ) > $sepDiv ) {
        my $sepsLength = ceil( @alphabet / $sepDiv );
        $sepsLength++ if $sepsLength == 1;
        if ( $sepsLength > @seps ) {
            push @seps => splice @alphabet, 0, $sepsLength - @seps;
        }
    }

    @alphabet = consistent_shuffle( \@alphabet, $self->salt );
    my $guardCount = ceil( @alphabet / $guardDiv );

    @guards
        = @alphabet < 3
        ? splice @seps, 0, $guardCount
        : splice @alphabet, 0, $guardCount;

    $self->_set_chars( \@alphabet );
    $self->_set_seps( \@seps );
    $self->_set_guards( \@guards );
}

sub encode_hex {
    my ( $self, $str ) = @_;

    return '' unless $str =~ /^[0-9a-fA-F]+$/;

    my @num;
    push @num, '1' . substr $str, 0, 11, '' while $str;

    @num = map { bignum(0)->from_hex($_) } @num;

    $self->encode(@num);
}

sub decode_hex {
    my ( $self, $hash ) = @_;

    my @res = $self->decode($hash);

    @res ? join '' => map { substr( bignum($_)->to_hex, 1 ) } @res : '';
}

sub encrypt {
    shift->encode(@_);
}

sub decrypt {
    shift->decode(shift);
}

sub encode {
    my ( $self, @num ) = @_;

    return '' unless @num;
    map { return '' unless defined and /^[0-9]+$/ } @num;

    my $num = [ map { bignum($_) } @num ];

    my @alphabet = @{ $self->chars };
    my @res;

    my $numHashInt = bignum(0);
    for my $i ( 0 .. $#$num ) {
        $numHashInt += $num->[$i] % ( $i + 100 );
    }

    my $lottery = $res[0] = $alphabet[ $numHashInt % @alphabet ];

    for my $i ( 0 .. $#$num ) {
        my $n = bignum( $num->[$i] );
        my @s = ( $lottery, split( // => $self->salt ), @alphabet )
            [ 0 .. @alphabet ];

        @alphabet = consistent_shuffle( \@alphabet, \@s );
        my $last = to_alphabet( $n, \@alphabet );

        push @res => split // => $last;

        if ( $i + 1 < @$num ) {
            $n %= ord($last) + $i;
            my $sepsIndex = $n % @{ $self->seps };
            push @res, $self->seps->[$sepsIndex];
        }
    }

    if ( @res < $self->minHashLength ) {
        my $guards     = $self->guards;
        my $guardIndex = ( $numHashInt + ord $res[0] ) % @$guards;
        my $guard      = $guards->[$guardIndex];

        unshift @res, $guard;

        if ( @res < $self->minHashLength ) {
            $guardIndex = ( $numHashInt + ord $res[2] ) % @$guards;
            $guard      = $guards->[$guardIndex];

            push @res, $guard;
        }
    }

    my $halfLength = int @alphabet / 2;
    while ( @res < $self->minHashLength ) {
        @alphabet = consistent_shuffle( \@alphabet, \@alphabet );
        @res = (
            @alphabet[ $halfLength .. $#alphabet ],
            @res, @alphabet[ 0 .. $halfLength - 1 ]
        );

        if ( ( my $excess = @res - $self->minHashLength ) > 0 ) {
            @res = splice @res, int $excess / 2, $self->minHashLength;
        }
    }

    join '' => @res;
}

sub decode {
    my ( $self, $hash ) = @_;

    return unless $hash;
    return unless defined wantarray;

    my $res  = [];
    my $orig = $hash;

    my $guard = join '|', map {quotemeta} @{ $self->guards };
    my @hash = grep { $_ ne '' } split /$guard/ => $hash;
    my $i = ( @hash == 3 || @hash == 2 ) ? 1 : 0;

    return unless defined( $hash = $hash[$i] );
    my $lottery = substr $hash, 0, 1;
    $hash = substr $hash, 1;

    my $sep = join '|', @{ $self->seps };
    @hash = grep { $_ ne '' } split /$sep/ => $hash;

    my @alphabet = @{ $self->chars };
    for my $part (@hash) {
        my @s = ( $lottery, split( // => $self->salt ), @alphabet )
            [ 0 .. @alphabet ];

        @alphabet = consistent_shuffle( \@alphabet, \@s );
        push @$res => from_alphabet( $part, \@alphabet );
    }

    return unless $self->Hashids::encode(@$res) eq $orig;

    wantarray ? @$res : @$res == 1 ? $res->[0] : $res;
}

1;
__END__

=encoding utf-8

=for stopwords minHashLength

=head1 NAME

Hashids - generate short hashes from numbers

=head1 SYNOPSIS

    use Hashids;
    my $hashids = Hashids->new('this is my salt');

    # encrypt a single number
    my $hash = $hashids->encode(123);          # 'YDx'
    my $number = $hashids->decode('YDx');      # 123

    # or a list
    $hash = $hashids->encode(1, 2, 3);         # 'laHquq'
    my @numbers = $hashids->decode('laHquq');  # (1, 2, 3)

    # also get results in an arrayref
    my $numbers = $hashids->decode('laHquq');  # [1, 2, 3]

=head1 DESCRIPTION

This is a port of the Hashids JavaScript library for Perl.

Hashids was designed for use in URL shortening, tracking stuff,
validating accounts or making pages private (through abstraction.)
Instead of showing items as C<1>, C<2>, or C<3>, you could show them as
C<b9iLXiAa>, C<EATedTBy>, and C<Aaco9cy5>.  Hashes depend on your salt
value.

B<IMPORTANT>: This implementation follows the v1.0.0 API release of
hashids.js.  An older API of hashids.js (v0.1.4) can be found in Hashids
version 0.08 and earlier releases; if you have code that depends on this
API version, please use a tool like L<Carton> to pin your Hashids
install to the older version.

This implementation is also compatible with the v0.3.x hashids.js API.

=head1 METHODS

=head2 new

    my $hashids = Hashids->new();

Make a new Hashids object.  This constructor accepts a few options:

    my $hashids = Hashids->new(
        salt          => 'this is my salt',
        alphabet      => 'abcdefghijklmnop',
        minHashLength => 8
    );

=over

=item  salt

Salt string, this should be unique per Hashids object.  Must be either
as long or shorter than the alphabet length, as a longer salt string
than the alphabet introduces false collisions.

=item  alphabet

Alphabet set to use.  This is optional as Hashids comes with a default
set suitable for URL shortening.  Should you choose to supply a custom
alphabet, make sure that it is at least 16 characters long, has no
spaces, and only has unique characters.

=item  minHashLength

Minimum hash length.  Use this to control how long the generated hash
string should be.

=back

You can also construct with just a single argument for the salt, leaving
the alphabet and minHashLength at their defaults:

    my $hashids = Hashids->new('this is my salt');

=head2 encode

    my $hash = $hashids->encode($x, [$y, $z, ...]);

Encode a single number (or a list of numbers) into a hash string.

=head2 encrypt

Alias for L</encode>, for compatibility with v0.3.x hashids.js API.

=head2 encode_hex

    my $hash = $hashids->encode_hex('deadbeef');

Encode a hex string into a hash string.

=head2 decode

    my $number = $hashids->decode($hash);

Decode a hash string into its number (or numbers.)  Returns either a
simple scalar if it is a single number, an arrayref of numbers if it
decrypted a set, or C<undef> if given bad input.  Use L<perlfunc/ref> on
the result to ensure proper usage.

You can also retrieve the result as a proper list by assigning it to an
array variable, by doing so you will always get a list of one or more
numbers that are decrypted from the hash, or the empty list if none were
found:

    my @numbers = $hashids->decode($hash);

=head2 decrypt

Alias for this L</decode>, for compatibility with v0.3.x hashids.js API.

=head2 decode_hex

    my $hex_string = $hashids->decode_hex($hash);

Opposite of L</encode_hex>.  Unlike L</decode>, this will always return
a string, including the empty string if the hash is invalid.

=head1 SEE ALSO

L<Hashids|http://www.hashids.org>

=head1 LICENSE

The MIT License (MIT)

Copyright (C) Zak B. Elep.

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

=head1 AUTHOR

Zak B. Elep E<lt>zakame@cpan.orgE<gt>

Original Hashids JavaScript library written by L<Ivan
Akimov|http://twitter.com/ivanakimov>

=head1 THANKS

Props to L<Jofell Gallardo|http://twitter.com/jofell> for pointing this
excellent project to me in the first place.

Many thanks to L<C. A. Church|https://github.com/thisdroneeatspeople>
and L<Troy Morehouse|https://github.com/tmorehouse> for their fixes and
updates.

=cut
